solr/solrj/src/java/org/apache/solr/common/util/JavaBinCodec.java

```diff
@@ -1080,6 +1083,15 @@ public boolean writePrimitive(Object val) throws IOException {
       daos.writeByte(DATE);
       daos.writeLong(((Date) val).getTime());
       return true;
+    } else if (val instanceof Instant) {
+      daos.writeByte(DATE);
+      daos.writeLong(((Instant) val).toEpochMilli());
+    } else if (val instanceof LocalDate) {
+      daos.writeByte(DATE);
+      daos.writeLong(((LocalDate) val).toEpochDay());
```

Comment:
I think EpochDay is the number of Days since epoch.

```suggestion
      daos.writeLong(((LocalDate) val).atStartOfDay(ZoneId.systemDefault()).toEpochSecond());
```

with the `ZoneId.systemDefault()`, you might be able to support `LocalDateTime` as well.

On the other hand, we don't necessarily know what zone is correct, and the system default might be wrong. Maybe we just remove the `LocalDate` support, like you did `LocalDateTime`.

Reply:
Thanks for spotting this and the proposed fix, but you got me thinking: Solr already has ParseDateFieldUpdateProcessorFactory to parse date strings and all JSR310 classes have a toString representation which can be parsed by it. Instead of trying to convert all to Date I've changed the code to convert them to their String representation and then ParseDateFieldUpdateProcessorFactory can parse it back on the server side. I think that's a better solution?

Reply:
Honestly I think it's better to use Date than String. Will probably revert those changes before moving forward. We would prefer this to work out of the box for most people without them having to add an UpdateProcessorFactory.